pennylane_lightning/src/Apply.cpp

```diff
@@ -156,56 +165,56 @@ void Pennylane::adjointJacobian(
             throw std::invalid_argument("The operation is not supported using the adjoint differentiation method");
         } else if ((operations[i] != "QubitStateVector") && (operations[i] != "BasisState")) {
             // copy |phi> to |mu> before applying Uj*
-            CplxType* phiCopyArr = new CplxType[phi.length];
-            std::memcpy(phiCopyArr, phi.arr, sizeof(CplxType)*phi.length);
-            Pennylane::StateVector mu(phiCopyArr, phi.length);
+            CplxType* phiCopyArr = new CplxType[lambdaState.length];
+            std::memcpy(phiCopyArr, lambdaState.arr, sizeof(CplxType)*lambdaState.length);
+            Pennylane::StateVector mu(phiCopyArr, lambdaState.length);
 
             // create |phi'> = Uj*|phi>
             Pennylane::constructAndApplyOperation(
-                phi,
+                lambdaState,
                 operations[i],
                 opWires[i],
                 opParams[i],
                 true,
-                opWires[i].size()
+                qubits
             );
 
-            if (std::find(trainableParams.begin(), trainableParams.end(), paramNumber) != trainableParams.end()) {
-                // create iH|phi> = d/d dUj/dtheta Uj* |phi> = dUj/dtheta|phi'>
-                unique_ptr<AbstractGate> gate = constructGate(operations[i], opParams[i]);
-                // double scalingFactor = gate->generatorScalingFactor;
-                double scalingFactor = Pennylane::RotationYGate::generatorScalingFactor;
-                Pennylane::applyGateGenerator(
-                    mu,
-                    gate,
-                    opWires[i],
-                    opWires[i].size()
-                );
 
-                for (unsigned int j = 0; j < lambdas.size(); j++) {
-                    int lambdaStateSize = lambdas[j].length;
+            // We have a parametrized gate
+            if (!opParams[i].empty()){
+                if (std::find(trainableParams.begin(), trainableParams.end(), current_param_idx) != trainableParams.end()) {
+
+                    // create iH|phi> = d/d dUj/dtheta Uj* |phi> = dUj/dtheta|phi'>
+                    unique_ptr<AbstractGate> gate = constructGate(operations[i], opParams[i]);
+                    // double scalingFactor = gate->generatorScalingFactor;
+                    double scalingFactor = Pennylane::RotationYGate::generatorScalingFactor;
```

Comment:
The scaling factor will have to be updated here to be specific to the operation (e.g., the phase shift operation will fail due to having another scaling factor than most parameterized gates).